doc/developer/design/20231213_compute_operator_hydration_status_tracking.md

```diff
@@ -0,0 +1,192 @@
+
+* **Efficiency:** Given the number of operators in a typical dataflow, the cost of tracking all their frontiers would be significant.
+  While we can't avoid some overhead from adding this feature, we would like to keep it as small as possible.
+* **Understandability:** Many of Compute's dataflow operators are implementation details that don't relate to the computed query (e.g. logging operators).
+  Showing information about these operators to users would at best introduce needless noise and at worst be actively confusing.
+* **Technical Feasibility:** Attaching hydration monitoring to every dataflow operator is difficult given the current code structure.
+  A good solution would likely require us to implement support into the Timely Dataflow runtime.
+
+Instead, this design proposes to only report hydration status of dataflow regions that correspond to LIR nodes.
+LIR is the low-level intermediate representation used by Materialize's optimizer.
+Users can show their queries as LIR plans using the `EXPLAIN PHYSICAL PLAN` syntax.
+LIR plans are the lowest level of query representation we document and that (advanced) users can therefore be expected to understand.
+LIR nodes map directly to groups of operators in a dataflow, so it is possible for us to collect hydration status information at this level.
+
+While it would be desirable to also provide hydration progress information at a higher level, we consider that outside the scope of this design.
+
+Reporting LIR-level hydration status requires a number of pieces that don't exist yet in our code:
+
+1. [Assigning IDs to LIR nodes during query planning](#lir-node-ids)
+2. [Creating a separate region for each LIR node during Compute rendering](#lir-dataflow-regions)
+3. [Collecting LIR region hydration status during dataflow processing](#hydration-status-logging)
+4. [Reporting LIR region hydration status to the controller](#reporting-introspection-data)
+
+While there is some amount of lifting required to implement all these pieces, our hope is that they will prove generally useful.
+(1) and (2) together allow mapping low-level dataflow information back to `EXPLAIN` plans and will be useful for making our internal memory visualizer user-facing.
+They are also a precondition for surfacing runtime information, like memory usage, in `EXPLAIN` plans.
+(4) gives us a reliable way to expose dataflow introspection information in SQL, which will be useful in alleviating some of the pain currently caused by the bad ergonomics of our per-replica introspection mechanism.
+
+### LIR Node IDs
+
+To make LIR nodes identifiable, we assign IDs to them (referred to as _LIDs_ in the following).
+LIDs need to be unique only within a dataflow, as they can be namespaced by the dataflow ID, or by the `GlobalId` of the SQL object that created the dataflow.
+
+This design proposes that LIDs should be increasing integers.
```

Comment:
Do you have an opinion where this association should be encoded (in-line with the `Plan` node variants, as a separate map, or something else).

Reply:
I think we have two choices:

* Give each `Plan` variant a `node_id` field.
* Rename the `Plan` enum to `PlanInner` and introduce a `struct Plan { id: NodeId, inner: PlanInner }`

The latter is cleaner because it avoids a bunch of repetition, but also looks to be like a huge amount of effort changing types everywhere that touches `Plan`s. So I propose doing the former unless there are strong opinions to the contrary.

> as a separate map

I'm not sure how to do that. What would the left side of such a map look like? We can't do `&Plan -> NodeId`, because that would make it impossible to move around the plan anymore.

Reply:
> The latter is cleaner because it avoids a bunch of repetition, but also looks to be like a huge amount of effort changing types everywhere that touches `Plan`s.

Actually I'll also give this a shot at implementing, the per-variant version turns also out to introduce a lot of churn.

Reply:
Rather then 

```rust
struct Plan { id: NodeId, inner: PlanInner }
```

can we have 

```rust
struct Plan<M = NodeId> { meta: M, inner: PlanInner }
```

Then we can extend the association between nodes and dataflow regions to MIR plans more easily by just changing this to

```rust
struct Plan<M = (LID, MID)> { meta: M, inner: PlanInner }
```

Reply:
I think this can be just as easily done with `struct Plan { id: NodeId, inner: PlanInner }` but using your preferred IDE's rename feature, or by changing the definition of `NodeId`. I'd like to not prematurely introduce generics unless there is a good reason to/we actually need them, as they just make the code harder to read otherwise.